### ISSUE TITLE
node.js child_process.fork, can not attach to child process.

### ISSUE THREAD

**Issue by Dzenly** [Microsoft/vscode#52322]
- VSCode Version: 1.24.0
- OS Version: Ubuntu 18, x64.
- Node.js: 8.11.3

There are so many posts about this trouble, but I can not find clear explanation or solutions among them.

I have such **parent.js**:
```js
'use strict';

const { fork } = require('child_process');
const path = require('path');

console.log('I am in parent');

// --inspect=0 means random free port.
fork(path.join(__dirname, 'child.js'), [], { execArgv : ['--inspect=0'] });

// To keep the parent process alive.
setTimeout(()=> console.log('Parent, After timeout'), 15000);
```
And such **child.js**:
```js
'use strict';
console.log('I am in child'); // set breakpoint here.
setTimeout(()=> console.log('Child, After timeout'), 15000);
```
**launch.json**
```
{
      "type": "node",
      "request": "launch",
      "name": "File",
      "program": "${file}",
      "console": "integratedTerminal",
      "autoAttachChildProcesses": false, // true changes nothing.
      "preLaunchTask": "clear",
      "protocol": "inspector",
      // "outputCapture": "console"
    },
```
**When I debugging this in WebStorm, everything is ok, I got my break point.**

When I debugging this in VS Code I get the error dialog box, 

```
Cannot connect to runtime process, timeout after 10000 ms - (reason: Cannot connect to the target: connect ECONNREFUSED 127.0.0.1:9229
```
Terminal:
```
/home/alexey/.nvm/versions/node/v8.11.3/bin/node --inspect-brk=29234 tests/parent.js
Debugger listening on ws://127.0.0.1:29234/258222ef-105f-4b0a-a3dc-59521d139afd
For help see https://nodejs.org/en/docs/inspector
Debugger attached.
I am in parent
Debugger listening on ws://127.0.0.1:46381/dbadea56-f572-40f7-8f09-f3950205056a
For help see https://nodejs.org/en/docs/inspector
I am in child
Parent, After timeout
Child, After timeout
Waiting for the debugger to disconnect...
Killed
```

Ok. Let's suppose that VS code can not detect port for child. So, I use
```js
fork(path.join(__dirname, 'child.js'), [], { execArgv : ['--inspect=9229'] });
```
in **parent.js**.

Now I have in terminal:
```
/home/alexey/.nvm/versions/node/v8.11.3/bin/node --inspect-brk=7533 tests/parent.js
Debugger listening on ws://127.0.0.1:7533/5043da39-bc59-4501-a9df-1629e88ae9a5
For help see https://nodejs.org/en/docs/inspector
Debugger attached.
I am in parent
Debugger listening on ws://127.0.0.1:9229/18470a55-f7e3-47c5-af75-b52c9fb5a8dd
For help see https://nodejs.org/en/docs/inspector
I am in child
Debugger attached.
Parent, After timeout
Child, After timeout
Waiting for the debugger to disconnect...
Waiting for the debugger to disconnect...
Killed
```

No errors, but breakpoints in child.js do not work.

So what is the problem core ? VS Code can not attach two processes simultaneously? It can not detect debug port of child processes?  Or what ?

**Comment by weinand** [comment]
You've see this explanation?
https://code.visualstudio.com/docs/nodejs/nodejs-debugging#_automatically-attach-debugger-to-nodejs-subprocesses

Passing a "0" is not a good way to let VS Code know what the real process id is.

**Comment by Dzenly** [comment]
Thank you for info.

Indeed, `explicit port` + `--inspect-brk instead of --inspect` + `"autoAttachChildProcesses": true` helps.
And indeed, `--inspect=9229` (no -brk) + break point after some timeout - work.

Interesting, that:
```js
fork(path.join(__dirname, 'child-no-git.js'), [], { execArgv : ['--inspect-brk=' + (process.debugPort + 1)] });
```
also works in VS Code.
So, seems like VS Code is specialized on following strings from cluster.js:
https://github.com/nodejs/node/blob/75c6d9dd95bbdb926d84cb3a9f518eff88651709/lib/cluster.js#L296-L319
Which adds 1 to find next free port for debug.

I don't understand your comment about 0 and process id. 0 means random free port for debugger.

Also I wondering why WebStorm does not have such limitations.
`--inspect=0` works in WebStorm.
So 
1. WebStorm does not require some delay to register breakpoints.
2. WebStorm detects random port even if it does not equal process.debugPort + 1.
Say, parent have debugPort 37531, child have random debugPort=41817 and breakpoint on console.log() works.

Can somebody explain these VS Code limitations ?

**Comment by weinand** [comment]
@Dzenly glad, that my suggestions were helpful.

Some additional comments:

you said:
> "VS Code is specialized on following strings from cluster.js"

No, we are not following cluster.js but we just analyse the (official) arguments passed to node.js and try to detect the used port from that. That explains why passing `--inspect=0` doesn't work because `0` is not the effective port (yes, I talked about "id" in my comment from above but really meant "port"). 

> "WebStorm does not have such limitations"

WebStorm most likely extracts the port from the console output: "Debugger listening on ws://127.0.0.1:9229/18470a55-f7e3-47c5-af75-b52c9fb5a8dd" (which fails in some other situations). VS Code uses a different approach that fails for the `0` case.

> "Interesting, that: --inspect-brk=' + (process.debugPort + 1) also works in VS Code.

I don't follow that statement. VS Code looks for the port argument passed to node.js. It doesn't care about whether this is a random port, or just the "next" (+1) port. They all work.

> "WebStorm does not require some delay to register breakpoints."

VS Code does not need a "delay". It just requires that node.js stops on the entry point so that the debugger gets a chance to register "early" breakpoints. For exactly that reason node.js supports the "--inspect-brk" argument. Without this it is impossible to hit early breakpoints. 
WebStorm has the same problem and they probably use "--inspect-brk" behind the scenes.

**Comment by Dzenly** [comment]
@weinand 
Thank you for your answers. Now I have understanding how it works in VS Code.

> WebStorm most likely extracts the port from the console output: "Debugger listening on ws://127.0.0.1:9229/18470a55-f7e3-47c5-af75-b52c9fb5a8dd" (which fails in some other situations). VS Code uses a different approach that fails for the 0 case.

Ok, I made a feature request.
https://github.com/Microsoft/vscode/issues/52373
Because it is inconvenient to use constants or find free port 'by hands'. And for now I did not meet any problems in WS.

> WebStorm has the same problem and they probably use "--inspect-brk" behind the scenes.
> Without this it is impossible to hit early breakpoints.

I made console.log(process.execArgv); in child process for `--inspect=0` and there is not any `-brk` in WebStorm. But breakpoints work perfectly.
So, seems like WebStorm does'not use `--inspect-brk` for child process. And works perfectly with `--inspect` without `-brk`. I will ask them how they do it.
